### PR TITLE
API diff between .NET 8 and .NET 9

### PR DESCRIPTION
**Pull request by carlossanlop**
Repo area owners:

- ASP.NET - @dotnet/aspnet-api-review 
- WinForms - @merriemcgaw @lonitra @JeremyKuhne
- WPF - @singhashish-wpf and @pchaurasia14
- Libraries - @artl93 @jeffschwMSFT @jeffhandley @karelz @ericstj @stephentoub @SamMonoRT 

Libraries area owners:
- System @dotnet/area-system-runtime
- System.Buffers @dotnet/area-system-buffers 
- System.CodeDom @dotnet/area-system-codedom 
- System.Collections @dotnet/area-system-collections 
- System.ComponentModel @dotnet/area-system-componentmodel 
- System.Data @SamMonoRT @ajcvickers @roji
- System.Diagnostics @dotnet/area-system-diagnostics 
- System.Diagnostics.Metrics @dotnet/area-system-diagnostics-metric 
- System.Drawing @dotnet/area-system-drawing @JeremyKuhne 
- System.Formats.Asn1 @dotnet/area-system-formats-asn1
- System.Formats.Nrbf @dotnet/area-system-formats-nrbf  
- System.Formats.Tar @dotnet/area-system-formats-tar 
- System.IO @dotnet/area-system-io 
- System.IO.Compression @dotnet/area-system-io-compression 
- System.IO.Pipelines @adityamandaleeka @davidfowl @halter73 
- System.IO.Pipes @dotnet/area-system-io 
- System.IO.Ports @dotnet/area-system-io-ports 
- System.Linq @dotnet/area-system-linq 
- System.Management @dotnet/area-system-management 
- System.Memory @dotnet/area-system-memory 
- System.Net @dotnet/ncl 
- System.Numerics @dotnet/area-system-numerics @dotnet/area-system-numerics-tensors 
- System.Reflection @dotnet/area-system-reflection @dotnet/area-system-reflection-emit @dotnet/area-system-reflection-metadata 
- System.Runtime.CompilerServices @dotnet/area-system-runtime-compilerservices 
- System.Runtime.InteropServices @dotnet/interop-contrib 
- System.Runtime.Intrinsics @dotnet/area-system-runtime-intrinsics 
- System.Security @dotnet/area-system-security 
- System.Text @dotnet/area-system-text-encoding 
- System.Text.Json @dotnet/area-system-text-json 
- System.Threading @dotnet/area-system-threading-channels @dotnet/area-system-threading-tasks @kouvel @mangod9 
- System.Timers @dotnet/area-system-runtime 

**Known api-diff tool issues**:

If yo usee any of these, please provide a GitHub suggestion in this PR to correct it: 

- The AsmDiff tool sometimes skips adding `{` or `{}` [arcade#10981](https://github.com/dotnet/arcade/issues/10981)(https://github.com/dotnet/arcade/issues/13814) 
- RequiresPreviewAttribute not getting auto added [arcade#14498](https://github.com/dotnet/arcade/issues/14498)
- Changes to `protected internal` visibility should stop showing. https://github.com/dotnet/arcade/issues/14579
- The AsmDiff tool shows APIs with decorating attribute changes as removed and re-added, instead of showing only the API. Please ignore this, it might be confusing but technically it's not incorrect: [arcade#13814]

### CODE REVIEW

**Review by roji — APPROVED** [review]
Approved for System.Data (SqlDbType.Json)

**Comment by bartonjs** [comment]
* System.Formats.Asn1 LGTM
* System.Security.* LGTM

File: release-notes/9.0/9.0.0/api-diff/Microsoft.NETCore.App/9.0.0_System.Diagnostics.md [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+-        public Activity StartActivity(string name, ActivityKind kind, ActivityContext parentContext, IEnumerable<KeyValuePair<string, object?>>? tags = null, IEnumerable<ActivityLink> links = null, DateTimeOffset startTime = default(DateTimeOffset));
++        public Activity? StartActivity(string name, ActivityKind kind, ActivityContext parentContext, IEnumerable<KeyValuePair<string, object?>>? tags = null, IEnumerable<ActivityLink>? links = null, DateTimeOffset startTime = default(DateTimeOffset));
+-        public Activity StartActivity(string name, ActivityKind kind, string parentId, IEnumerable<KeyValuePair<string, object?>>? tags = null, IEnumerable<ActivityLink> links = null, DateTimeOffset startTime = default(DateTimeOffset));
++        public Activity? StartActivity(string name, ActivityKind kind, string? parentId, IEnumerable<KeyValuePair<string, object?>>? tags = null, IEnumerable<ActivityLink>? links = null, DateTimeOffset startTime = default(DateTimeOffset));
```
Comment (tarekgh) [inline release-notes/9.0/9.0.0/api-diff/Microsoft.NETCore.App/9.0.0_System.Diagnostics.md:27]: These APIs were already using nullable types so I think this is not right diff.

File: release-notes/9.0/9.0.0/api-diff/Microsoft.NETCore.App/9.0.0_System.Diagnostics.Metrics.md [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+     }
++    public sealed class InstrumentAdvice<T> where T : struct {
++        public InstrumentAdvice();
++        public IReadOnlyList<T>? HistogramBucketBoundaries { get; set; }
++    }
+     public readonly struct Measurement<T> where T : struct {
++        public Measurement(T value, in TagList tags);
+     }
+-    public delegate void MeasurementCallback<T>(Instrument instrument, T measurement, ReadOnlySpan<KeyValuePair<string, object?>> tags, object? state);
++    public delegate void MeasurementCallback<T>(Instrument instrument, T measurement, ReadOnlySpan<KeyValuePair<string, object?>> tags, object? state) where T : struct;
```
Comment (tarekgh) [inline release-notes/9.0/9.0.0/api-diff/Microsoft.NETCore.App/9.0.0_System.Diagnostics.Metrics.md:32]: I don't think this has changed 

[https://github.com/dotnet/runtime/blob/c1ae962610643a3c17bcf9621017ac42ff7703f7/src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/MeterListener.cs#L12](https://github.com/dotnet/runtime/blob/c1ae962610643a3c17bcf9621017ac42ff7703f7/src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/MeterListener.cs#L12)

File: release-notes/9.0/9.0.0/api-diff/Microsoft.NETCore.App/9.0.0_System.Diagnostics.Metrics.md [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+     public class Meter : IDisposable {
++        public Gauge<T> CreateGauge<T>(string name) where T : struct;
++        public Gauge<T> CreateGauge<T>(string name, string unit = null, string description = null, IEnumerable<KeyValuePair<string, object?>>? tags = null) where T : struct;
++        public Histogram<T> CreateHistogram<T>(string name) where T : struct;
+-        public Histogram<T> CreateHistogram<T>(string name, string unit = null, string description = null) where T : struct;
++        public Histogram<T> CreateHistogram<T>(string name, string unit, string description) where T : struct;
+-        public Histogram<T> CreateHistogram<T>(string name, string unit, string description, IEnumerable<KeyValuePair<string, object?>> tags) where T : struct;
++        public Histogram<T> CreateHistogram<T>(string name, string unit, string description, IEnumerable<KeyValuePair<string, object?>>? tags) where T : struct;
```
Comment (tarekgh) [inline release-notes/9.0/9.0.0/api-diff/Microsoft.NETCore.App/9.0.0_System.Diagnostics.Metrics.md:40]: This didn't change either:

[https://github.com/dotnet/runtime/blob/c1ae962610643a3c17bcf9621017ac42ff7703f7/src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/Meter.cs#L166](https://github.com/dotnet/runtime/blob/c1ae962610643a3c17bcf9621017ac42ff7703f7/src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/Meter.cs#L166)

**Comment by tarekgh** [comment]
I added some comments, LGTM otherwise:

- System.Diagnostics (Activity and ActivitySource)
- System.Diagnostics.Meterics
- System.Environment.CpuUsage/ProcessCpuUsage
- System.TimeSpan

**Review by eiriktsarpalis — APPROVED** [review]
Linq, collections and STJ LGTM

**Review by tannergooding — APPROVED** [review]

File: release-notes/9.0/9.0.0/api-diff/Microsoft.NETCore.App/9.0.0_System.Collections.Generic.md [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+     public class PriorityQueue<TElement, TPriority> {
++        public bool Remove(TElement element, out TElement removedElement, out TPriority priority, IEqualityComparer<TElement>? equalityComparer = null);
+         public sealed class UnorderedItemsCollection : ICollection, IEnumerable, IEnumerable<(TElement, TPriority)>, IReadOnlyCollection<(TElement, TPriority)> {
+-            IEnumerator<(TElement Element, TPriority Priority)> IEnumerable<(TElement, TPriority)>.GetEnumerator();
+
++            IEnumerator<(TElement Element, TPriority Priority)> IEnumerable<(TElement, TPriority)>.GetEnumerator();
+             public struct Enumerator : IDisposable, IEnumerator, IEnumerator<(TElement, TPriority)> {
+-                (TElement Element, TPriority Priority) IEnumerator<(TElement, TPriority)>.Current { get; }
+
++                (TElement Element, TPriority Priority) IEnumerator<(TElement, TPriority)>.Current { get; }
```
Comment (xparadoxical) [inline release-notes/9.0/9.0.0/api-diff/Microsoft.NETCore.App/9.0.0_System.Collections.Generic.md:200]: nothing seems to have changed here, and note the empty lines not present anywhere else (for some reason which could be related to this zero diff)

File: release-notes/9.0/9.0.0/api-diff/Microsoft.NETCore.App/9.0.0_System.Diagnostics.md (hunk 1 above)
Comment (carlossanlop) [reply release-notes/9.0/9.0.0/api-diff/Microsoft.NETCore.App/9.0.0_System.Diagnostics.md:27]: Weird. I see it. All the CreateActivity and StartActivity methods were already using the exact same nullability in return values and parameters. The Version property too. I'll remove them all from the diff.

https://github.com/dotnet/runtime/tree/release/8.0/src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/ActivitySource.cs#L69-L160

**Pull request merged**